src/EditorFeatures/VisualBasic/AutomaticEndConstructCorrection/AbstractCorrector.vb

```diff
@@ -0,0 +1,209 @@
+Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
+Imports Microsoft.VisualStudio.Text
+
+Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.AutomaticEndConstructCorrection
+    Partial Friend MustInherit Class AbstractCorrector
+        Implements ICorrector
+
+        Private ReadOnly _buffer As ITextBuffer
+        Private ReadOnly _waitIndicator As IWaitIndicator
+        Private ReadOnly _session As Session
+
+        Private _previousDocument As Document
+        Private _referencingViews As Integer
+
+        ' Private ReadOnly _session As Session
+
+        Protected Sub New(subjectBuffer As ITextBuffer, waitIndicator As IWaitIndicator)
+            _buffer = subjectBuffer
+            _waitIndicator = waitIndicator
+
+            _session = New Session(subjectBuffer, AddressOf ShouldReplaceText)
+            _previousDocument = Nothing
+            _referencingViews = 0
+        End Sub
+
+        Protected MustOverride Function IsAllowableWordAtIndex(lineText As String, wordStartIndex As Integer, wordLength As Integer) As Boolean
+        Protected MustOverride Function TryGetValidTokens(wordStartIndex As Integer, ByRef startToken As SyntaxToken, ByRef endToken As SyntaxToken, cancellationToken As CancellationToken) As Boolean
+        Protected MustOverride Function ShouldReplaceText(text As String) As Boolean
+
+        Protected ReadOnly Property PreviousDocument As Document
+            Get
+                Return _previousDocument
+            End Get
+        End Property
+
+        Public Sub Connect() Implements ICorrector.Connect
+            If _referencingViews = 0 Then
+                AddHandler _buffer.Changing, AddressOf OnTextBufferChanging
+                AddHandler _buffer.Changed, AddressOf OnTextBufferChanged
+            End If
+
+            _referencingViews = _referencingViews + 1
+        End Sub
+
+        Public Sub Disconnect() Implements ICorrector.Disconnect
+            If _referencingViews = 1 Then
+                RemoveHandler _buffer.Changed, AddressOf OnTextBufferChanged
+                RemoveHandler _buffer.Changing, AddressOf OnTextBufferChanging
+            End If
+
+            _referencingViews = Math.Max(_referencingViews - 1, 0)
+        End Sub
+
+        Public ReadOnly Property IsDisconnected As Boolean Implements ICorrector.IsDisconnected
+            Get
+                Return _referencingViews = 0
+            End Get
+        End Property
+
+        Protected Function CreateTrackingSpans(startToken As SyntaxToken, endToken As SyntaxToken, snapshot As ITextSnapshot) As IEnumerable(Of ITrackingSpan)
+            Return {CreateTrackingSpan(startToken, snapshot), New LetterOnlyTrackingSpan(endToken.Span.ToSnapshotSpan(snapshot))}
+        End Function
+
+        Private Function CreateTrackingSpan(token As SyntaxToken, snapshot As ITextSnapshot) As ITrackingSpan
+            Return New LetterOnlyTrackingSpan(token.Span.ToSnapshotSpan(snapshot))
```

Comment:
Note that XML tag names can include : for namespaces. If this is truly "letter-only" (I'm still trying to dig to find the list), then this will need expanding. Shouldn't be hard though.